keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,903 @@
+
+### Risks and Mitigations
+
+TBD
+
+## Design Details
+
+### Test Plan
+
+TBD
+
+### Upgrade / Downgrade Strategy
+
+TBD
+
+### Version Skew Strategy
+
+TBD
+
+## Implementation History
+
+## Drawbacks
+
+Why should this KEP _not_ be implemented - TBD.
+
+## Alternatives
+
+### CSIDriver.Status
+
+Alternatively, a `CSIDriver.Status` could combine all information in
+one object in a way that is both human-readable (albeit potentially
+large) and matches the lookup pattern of the scheduler. Updates could
+be done efficiently via `PATCH` operations. Finding information about
+all pools at once would be simpler.
+
+However, continuously watching this single object and retrieving
+information about just one pool would become more expensive. Because
+the API needs to be flexible enough to also support this for future
+use cases, this approach has been rejected.
+
+```
+type CSIDriver struct {
+    ...
+
+	// Specification of the CSI Driver.
+	Spec CSIDriverSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status of the CSI Driver.
+	// +optional
+	Status CSIDriverStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+type CSIDriverSpec struct {
+    ...
+
+	// CapacityTracking defines whether the driver deployment will provide
+	// capacity information as part of the driver status.
+	// +optional
+	CapacityTracking *bool `json:"capacityTracking,omitempty" protobuf:"bytes,4,opt,name=capacityTracking"`
+}
+
+// CSIDriverStatus represents dynamic information about the driver and
+// the storage provided by it, like for example current capacity.
+type CSIDriverStatus struct {
+	// Each driver can provide access to different storage pools
+	// (= subsets of the overall storage with certain shared
+	// attributes).
+	//
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	Storage []CSIStoragePool `patchStrategy:"merge" patchMergeKey:"name" json:"storage,omitempty" protobuf:"bytes,1,opt,name=storage"`
+}
+
+// CSIStoragePool identifies one particular storage pool and
+// stores the corresponding attributes.
+//
+// A pool might only be accessible from a subset of the nodes in the
+// cluster. That subset can be identified either via NodeTopology or
+// Nodes, but not both. If neither is set, the pool is assumed
+// to be available in the entire cluster.
+type CSIStoragePool struct {
+	// The name is some user-friendly identifier for this entry.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// NodeTopology can be used to describe a storage pool that is available
+	// only for nodes matching certain criteria.
+	// +optional
+	NodeTopology *v1.NodeSelector `json:"nodeTopology,omitempty" protobuf:"bytes,2,opt,name=nodeTopology"`
+
+	// Nodes can be used to describe a storage pool that is available
+	// only for certain nodes in the cluster.
+	//
+	// +listType=set
+	// +optional
+	Nodes []string `json:"nodes,omitempty" protobuf:"bytes,3,opt,name=nodes"`
```

Comment:
IMHO Nodes/NodeTopology should be in the status, because in networked storage environments it is possible for a storage array (StoragePool) to be healthy, but not reachable by a particular node (e.g. due to network path issues), and so making this dynamic seems like a prudent API design even if we don't take immediate advantage of it.

Reply:
Makes sense to me. The assumption was that it is immutable (node-local storage) or can be specified in fixed way (network attached storage with node selector), but I agree that it is better to be more flexible. Will change it.

Reply:
This also also removes the requirement that the spec must be unique because now the spec is just the driver name. The external-provisioner still needs to ensure that it doesn't produce more objects than really needed, but that now becomes an implementation detail of that component instead of something required by the API.